### PR TITLE
Ability to cache mark and index files for S3 disk

### PR DESCRIPTION
**Pull request by Jokser**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added cache layer for DiskS3 (cache to local disk mark and index files).

Detailed description / Documentation draft:
Introduced cache decorator over an IDisk.
Cache decorator tries to cache files by given predicated to a provided local disk.
writeFile() first writes a file to cache and after writing is finished copies this file to underlying disk.
readFile() tries to read from a cache first. If the file doesn't exist it initiates downloading this file from underlying disk to cache and when downloading is finished - returns read buffer from the cache.
Files moving, renaming, deleting are preserved in cache layer as well to prevent cache pollution with stale files.

### CODE REVIEW

File: src/Disks/DiskCacheWrapper.cpp [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+{
+}
+
+std::shared_ptr<FileDownloadMetadata> DiskCacheWrapper::acquireDownloadMetadata(const String & path) const
+{
+    std::unique_lock<std::mutex> lock{mutex};
+
+    auto it = file_downloads.find(path);
+    if (it != file_downloads.end() && !it->second.expired())
+        return it->second.lock();
```
Comment (akuzm) [inline src/Disks/DiskCacheWrapper.cpp:21]: `lock()` can return a nullptr when the object is deleted, looks like you don't check for that

Comment (Jokser) [reply src/Disks/DiskCacheWrapper.cpp:21]: I checked it by `expired()` call

File: tests/integration/test_merge_tree_s3_with_cache/test.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
+import logging
```
Comment (alesapin) [inline tests/integration/test_merge_tree_s3_with_cache/test.py:1]: Need to add `__init__.py` file in the same folder.

**Review by akuzm — APPROVED** [review]

**Comment by akuzm** [comment]
00933_test_fix_extra_seek_on_compressed_cache is in CI Fest #13076 
fuzzer failure is https://github.com/ClickHouse/ClickHouse/issues/13392

**Pull request merged**